core/src/main/java/tech/tablesaw/io/ReadOptions.java

```diff
@@ -327,8 +358,133 @@ public Builder minimizeColumnSizes() {
+      this.columnTypeFunction = columnTypeFunction;
+      return this;
+    }
+
+    /**
+     * Provide a map that determines ColumnType for given column names. Types for not present column
+     * names will be autodetected. To provide type for all column names use {@link
+     * #columnTypes(Function)} that generally is more efficient because it skips column type
+     * detection
+     */
+    public Builder columnTypesPartial(Map<String, ColumnType> columnTypeByName) {
+      if (columnTypeOptionsAlreadySet()) {
+        throw new IllegalStateException("columnTypes already set");
+      }
+      if (columnTypeByName != null) {
+        this.columnTypeMap = columnTypeByName;
+      }
+      return this;
+    }
+
+    private boolean columnTypeOptionsAlreadySet() {
+      return columnTypes != null
+          || columnTypeFunction != null
+          || completeColumnTypeFunction != null
+          || !columnTypeMap.isEmpty();
+    }
+
     public ReadOptions build() {
       return new ReadOptions(this);
     }
   }
+
+  /**
+   * Allows user to set column types. It can work in three ways:
+   *
+   * <ul>
+   *   <li>If no information is provided column types are autodetected
+   *   <li>A complete list of columns can be provided using {@link
+   *       ReadOptions.Builder#columnTypes(ColumnType[])} or {@link
+   *       ReadOptions.Builder#columnTypes(Function)}. This skips column type detection.
+   *   <li>Provide values for some column names using {@link
+   *       ReadOptions.Builder#columnTypesPartial(Map)} or {@link
+   *       ReadOptions.Builder#columnTypesPartial(Function)} (String, ColumnType)}. In this case
+   *       provided columnTypes are used and the others are autodetected
+   * </ul>
+   */
+  public static class ColumnTypeReadOptions {
+    final ColumnType[] columnTypesByIdx;
+    final Map<String, ColumnType> columnTypesByNameMap;
+    final Function<String, Optional<ColumnType>> columnTypesByNameFunction;
+    final Function<String, ColumnType> completeColumnTypesByNameFunction;
+
+    public static ColumnTypeReadOptions of(ColumnType[] allColumnTypes) {
+      return new ColumnTypeReadOptions(allColumnTypes, Collections.emptyMap(), null, null);
+    }
+
+    ColumnTypeReadOptions(
+        ColumnType[] columnTypesByIdx,
+        Map<String, ColumnType> columnTypesByNameMap,
+        Function<String, ColumnType> completeColumnTypesByNameFunction,
+        Function<String, Optional<ColumnType>> columnTypesByNameFunction) {
+      this.columnTypesByIdx = columnTypesByIdx;
+      this.columnTypesByNameMap = columnTypesByNameMap;
+      this.columnTypesByNameFunction = columnTypesByNameFunction;
+      this.completeColumnTypesByNameFunction = completeColumnTypesByNameFunction;
+    }
+
+    public Optional<ColumnType> columnType(int columnNumber, String columnName) {
+      if (columnTypesByIdx != null && columnNumber < columnTypesByIdx.length) {
+        return Optional.ofNullable(columnTypesByIdx[columnNumber]);
```

Comment:
If columnTypesByIdx, i don't see why this returns ofNullable. Please explain if you think this is correct.